README.md
# Azure Distributed Data Engineering Toolkit (AZTK)
Azure Distributed Data Engineering Toolkit (AZTK) is a python CLI application for provisioning on-demand Spark on Docker clusters in Azure. It's a cheap and easy way to get up and running with a Spark cluster, and a great tool for Spark users who want to experiment and start testing at scale.

This toolkit is built on top of Azure Batch but does not require any Azure Batch knowledge to use.

## Notable Features
- Spark cluster provision time of 5 minutes on average
- Spark clusters run in Docker containers
- Run Spark on a GPU enabled cluster
- Users can bring their own Docker image
- Ability to use low-priority VMs for an 80% discount
- Mixed Mode clusters that use both low-priority and dedicated VMs
- Built in support for Azure Blob Storage and Azure Data Lake connection
- [Tailored pythonic experience with PySpark, Jupyter, and Anaconda](https://github.com/Azure/aztk/wiki/PySpark-on-Azure-with-AZTK)
- [Tailored R experience with SparklyR, RStudio-Server, and Tidyverse](https://github.com/Azure/aztk/wiki/SparklyR-on-Azure-with-AZTK)
- Ability to run _spark submit_ directly from your local machine's CLI

## Setup

1. Install `aztk` with pip:
```bash
    pip install aztk
```
2. Initialize the project in a directory. This will automatically create a *.aztk* folder with config files in your working directory:
```bash
    aztk spark init
```
3. Login or register for an [Azure Account](https://azure.microsoft.com), navigate to [Azure Cloud Shell](https://shell.azure.com), and run:
```sh
wget -q https://raw.githubusercontent.com/Azure/aztk/v0.10.3/account_setup.sh -O account_setup.sh &&
chmod 755 account_setup.sh &&
/bin/bash account_setup.sh
```

4. Follow the on screen prompts to create the necessary Azure resources and copy the output into your `.aztk/secrets.yaml` file. For more information see [Getting Started Scripts](./01-Getting-Started-Script).


## Quickstart Guide

The core experience of this package is centered around a few commands.

```sh
# create your cluster
aztk spark cluster create
aztk spark cluster add-user
```
```sh
# monitor and manage your clusters
aztk spark cluster get
aztk spark cluster list
aztk spark cluster delete
```
```sh
# login and submit applications to your cluster
aztk spark cluster ssh
aztk spark cluster submit
```

### 1. Create and setup your cluster

First, create your cluster:
```bash
aztk spark cluster create --id my_cluster --size 5 --vm-size standard_d2_v2
```
- See our available VM sizes [here.](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/sizes) 
- The `--vm-size` argument must be the official SKU name which usually come in the form: "standard_d2_v2"
- You can create [low-priority VMs](https://docs.microsoft.com/en-us/azure/batch/batch-low-pri-vms) at an 80% discount by using `--size-low-pri` instead of `--size`
- By default, AZTK runs Spark 2.2.0 on an Ubuntu16.04 Docker image. More info [here](/docker-image)
- By default, AZTK will create a user (with the username **spark**) for your cluster
- The cluster id (`--id`) can only contain alphanumeric characters including hyphens and underscores, and cannot contain more than 64 characters.
- By default, you cannot create clusters of more than 20 cores in total. Visit [this page](https://docs.microsoft.com/en-us/azure/batch/batch-quota-limit#view-batch-quotas) to request a core quota increase.

More information regarding using a cluster can be found in the [cluster documentation](./docs/10-clusters.md)

### 2. Check on your cluster status
To check your cluster status, use the `get` command:
```bash
aztk spark cluster get --id my_cluster
```

### 3. Submit a Spark job

When your cluster is ready, you can submit jobs from your local machine to run against the cluster. The output of the spark-submit will be streamed to your local console. Run this command from the cloned AZTK repo:
```bash
// submit a java application
aztk spark cluster submit \
    --id my_cluster \
    --name my_java_job \
    --class org.apache.spark.examples.SparkPi \
    --executor-memory 20G \
    path\to\examples.jar 1000
    
// submit a python application
aztk spark cluster submit \
    --id my_cluster \
    --name my_python_job \
    --executor-memory 20G \
    path\to\pi.py 1000
```
- The `aztk spark cluster submit` command takes the same parameters as the standard [`spark-submit` command](https://spark.apache.org/docs/latest/submitting-applications.html), except instead of specifying `--master`, AZTK requires that you specify your cluster `--id` and a unique job `--name`
- The job name, `--name`, argument must be at least 3 characters long
    - It can only contain alphanumeric characters including hyphens but excluding underscores
    - It cannot contain uppercase letters
- Each job you submit **must** have a unique name
- Use the `--no-wait` option for your command to return immediately

Learn more about the spark submit command [here](./docs/20-spark-submit.md)

### 4. Log in and Interact with your Spark Cluster
Most users will want to work interactively with their Spark clusters. With the `aztk spark cluster ssh` command, you can SSH into the cluster's master node. This command also helps you port-forward your Spark Web UI and Spark Jobs UI to your local machine:
```bash
aztk spark cluster ssh --id my_cluster --user spark
```
By default, we port forward the Spark Web UI to *localhost:8080*, Spark Jobs UI to *localhost:4040*, and the Spark History Server to *localhost:18080*.

You can configure these settings in the *.aztk/ssh.yaml* file.

NOTE: When working interactively, you may want to use tools like Jupyter or RStudio-Server. To do so, you need to setup your cluster with the appropriate docker image and plugin. See [Plugins](./docs/15-plugins.md) for more information.

### 5. Manage and Monitor your Spark Cluster

You can also see your clusters from the CLI:
```
aztk spark cluster list
```

And get the state of any specified cluster:
```
aztk spark cluster get --id <my_cluster_id>
```

Finally, you can delete any specified cluster:
```
aztk spark cluster delete --id <my_cluster_id>
```

## FAQs
- [How do I connect to Azure Storage (WASB)?](./docs/30-cloud-storage.md)
- [I want to use a different version of Spark](./docs/12-docker-image.md)
- [How do I SSH into my Spark cluster's master node?](./docs/10-clusters.md#ssh-and-port-forwarding)
- [How do I interact with my Spark cluster using a password instead of an SSH-key?](./docs/10-clusters.md#interactive-mode)
- [How do I change my cluster default settings?](./docs/13-configuration.md)
- [How do I modify my *spark-env.sh*, *spark-defaults.conf* or *core-site.xml* files?](./docs/13-configuration.md)
- [How do I use GPUs with AZTK](./docs/60-gpu.md)
- [I'm a python user and want to use PySpark, Jupyter, Anaconda packages, and have a Pythonic experience.](https://github.com/Azure/aztk/wiki/PySpark-on-Azure-with-AZTK)
- [I'm a R user and want to use SparklyR, RStudio, Tidyverse packages, and have an R experience.](https://github.com/Azure/aztk/wiki/SparklyR-on-Azure-with-AZTK)

## Next Steps
You can find more documentation [here](./docs)


CHANGELOG.md
# Changelog

##  0.10.3 (2020-02-13)

**Bug Fixes**
* Pin pipenv dependencies in setup_host.sh (#709) ([5be9e87](https://github.com/Azure/aztk/commit/5be9e87)), closes [#710](https://github.com/Azure/aztk/issues/710)


##  0.10.2 (2018-12-07)

**Bug Fixes**
* Fix: Storage table never deleted (#690) ([723995c](https://github.com/Azure/aztk/commit/723995c)), closes [#690](https://github.com/Azure/aztk/issues/690)

**Internal Changes**
* Internal: Migrate off preview vsts queues (#685) ([a91690b](https://github.com/Azure/aztk/commit/a91690b)), closes [#685](https://github.com/Azure/aztk/issues/685)


##  0.10.1 (2018-11-02)

**Features**
* Include cluster creation time in cluster list and cluster get output (#678) ([a0bc2f0](https://github.com/Azure/aztk/commit/a0bc2f0)), closes [#678](https://github.com/Azure/aztk/issues/678)

**Bug Fixes**
* Verify Batch task has node_info (#681) ([385040d](https://github.com/Azure/aztk/commit/385040d)), closes [#681](https://github.com/Azure/aztk/issues/681)


##  0.10.0 (2018-10-29)

**Breaking Changes**
* Remove deprecated SDK API code (#671) ([fc50536](https://github.com/Azure/aztk/commit/fc50536)), closes [#671](https://github.com/Azure/aztk/issues/671)
* Remove custom scripts (#673) ([9e32b4b](https://github.com/Azure/aztk/commit/9e32b4b)), closes [#673](https://github.com/Azure/aztk/issues/673)
* Replaced states with Enums for ClusterState, JobState, ApplicationState (#677) ([e486536](https://github.com/Azure/aztk/commit/e486536)), closes [#677](https://github.com/Azure/aztk/issues/677)

**Features**
* Spark retry docker pull (#672) ([18b74e4](https://github.com/Azure/aztk/commit/18b74e4)), closes [#672](https://github.com/Azure/aztk/issues/672)
* Spark scheduling target (#661) ([4408c4f](https://github.com/Azure/aztk/commit/4408c4f)), closes [#661](https://github.com/Azure/aztk/issues/661)
* Spark submit scheduling internal (#674) ([8c2bf0c](https://github.com/Azure/aztk/commit/8c2bf0c)), closes [#674](https://github.com/Azure/aztk/issues/674)


##  0.9.1 (2018-10-5)
**Bug Fixes**
* Fix: pin all node dependencies not in Pipfile (#667) ([0606598](https://github.com/Azure/aztk/commit/0606598)), closes [#667](https://github.com/Azure/aztk/issues/667)
* Fix: vsts integration tests block (#657) ([4a60c8a](https://github.com/Azure/aztk/commit/4a60c8a)), closes [#657](https://github.com/Azure/aztk/issues/657)
* Fix: vsts mutliline secrets (#668) ([cb62207](https://github.com/Azure/aztk/commit/cb62207)), closes [#668](https://github.com/Azure/aztk/issues/668)


##  0.9.0 (2018-08-30)

**Breaking Changes**
* spark roll back scheduling disable (#653) ([93615d9](https://github.com/Azure/aztk/commit/93615d9)), closes [#653](https://github.com/Azure/aztk/issues/653)
* remove custom scripts (#650) ([442228a](https://github.com/Azure/aztk/commit/442228a)), closes [#650](https://github.com/Azure/aztk/issues/650)
* 0.9.0 deprecated code removal (#645) ([eef36dc](https://github.com/Azure/aztk/commit/eef36dc)), closes [#645](https://github.com/Azure/aztk/issues/645)
* SDK refactor (#622) ([b18eb69](https://github.com/Azure/aztk/commit/b18eb69)), closes [#622](https://github.com/Azure/aztk/issues/622)

**Features**
* Add ability to specify docker run options in toolkit config (#613) ([9d554c3](https://github.com/Azure/aztk/commit/9d554c3)), closes [#613](https://github.com/Azure/aztk/issues/613) [#3](https://github.com/Azure/aztk/issues/3)
* add brief flag to debug tool (#634) ([b7bdd8c](https://github.com/Azure/aztk/commit/b7bdd8c)), closes [#634](https://github.com/Azure/aztk/issues/634)
* first run docs update (#644) ([9098533](https://github.com/Azure/aztk/commit/9098533)), closes [#644](https://github.com/Azure/aztk/issues/644)
* SDK refactor (#622) ([b18eb69](https://github.com/Azure/aztk/commit/b18eb69)), closes [#622](https://github.com/Azure/aztk/issues/622)

**Bug Fixes**
* diagnostics function write error result bug (#649) ([293f297](https://github.com/Azure/aztk/commit/293f297)), closes [#649](https://github.com/Azure/aztk/issues/649)
* expose get cluster configuration API (#648) ([7c14648](https://github.com/Azure/aztk/commit/7c14648)), closes [#648](https://github.com/Azure/aztk/issues/648)
* remove bad node scripts import (#652) ([0a9ce94](https://github.com/Azure/aztk/commit/0a9ce94)), closes [#652](https://github.com/Azure/aztk/issues/652)
* typo in vsts build (#654) ([7c37b06](https://github.com/Azure/aztk/commit/7c37b06)), closes [#654](https://github.com/Azure/aztk/issues/654)
* update incompatible dependencies in setup.py (#639) ([f98d037](https://github.com/Azure/aztk/commit/f98d037)), closes [#639](https://github.com/Azure/aztk/issues/639)

**Internal Changes**
* fix pylint warnings (#651) ([828162e](https://github.com/Azure/aztk/commit/828162e)), closes [#651](https://github.com/Azure/aztk/issues/651)
* update vsts build (#635) ([6eda21e](https://github.com/Azure/aztk/commit/6eda21e)), closes [#635](https://github.com/Azure/aztk/issues/635)
* verify code formatting in build (#633) ([7730c46](https://github.com/Azure/aztk/commit/7730c46)), closes [#633](https://github.com/Azure/aztk/issues/633)


##  0.8.1 (2018-06-20)

**Bug Fixes**
* docs links version (#614) ([a8f8e92](https://github.com/Azure/aztk/commit/a8f8e92)), closes [#614](https://github.com/Azure/aztk/issues/614)
* set defaults for SparkConfiguration, add tests (#606) ([5306a2a](https://github.com/Azure/aztk/commit/5306a2a)), closes [#606](https://github.com/Azure/aztk/issues/606)
* spark debug tool filter out .venv, make debug tool testable (#612) ([4e0b1ec](https://github.com/Azure/aztk/commit/4e0b1ec)), closes [#612](https://github.com/Azure/aztk/issues/612)
* Suppress msrest warnings (#611) ([883980d](https://github.com/Azure/aztk/commit/883980d)), closes [#611](https://github.com/Azure/aztk/issues/611)


##  0.8.0 (2018-06-12)

**Deprecated Features**
* ClusterConfiguration fields vm_count and vm_count_low_pri have been renamed to size and size_low_priority

cli/config.py
import os
import yaml
from cli import log
import aztk.spark
from aztk.spark.models import (
    SecretsConfiguration,
    ServicePrincipalConfiguration,
    SharedKeyConfiguration,
    DockerConfiguration,
    ClusterConfiguration,
    UserConfiguration,
    PluginConfiguration,
)
from aztk.models.plugins.internal import PluginReference

def load_aztk_secrets() -> SecretsConfiguration:
    """
    Loads aztk from .aztk/secrets.yaml files(local and global)
    """
    secrets = SecretsConfiguration()
    # read global ~/secrets.yaml
    global_config = _load_secrets_config(
        os.path.join(aztk.utils.constants.HOME_DIRECTORY_PATH, '.aztk',
                     'secrets.yaml'))
    # read current working directory secrets.yaml
    local_config = _load_secrets_config()

    if not global_config and not local_config:
        raise "There is no secrets.yaml in either ./.aztk/secrets.yaml or .aztk/secrets.yaml"

    if global_config:  # GLobal config is optional
        _merge_secrets_dict(secrets, global_config)
    if local_config:
        _merge_secrets_dict(secrets, local_config)

    # Validate and raise error if any
    secrets.validate()
    return secrets


def _load_secrets_config(
        path: str = aztk.utils.constants.DEFAULT_SECRETS_PATH):
    """
        Loads the secrets.yaml file in the .aztk directory
    """
    if not os.path.isfile(path):
        return None

    with open(path, 'r') as stream:
        try:
            return yaml.load(stream)
        except yaml.YAMLError as err:
            raise aztk.error.AztkError(
                "Error in secrets.yaml: {0}".format(err))


def _merge_secrets_dict(secrets: SecretsConfiguration, secrets_config):
    service_principal_config = secrets_config.get('service_principal')
    if service_principal_config:
        secrets.service_principal = ServicePrincipalConfiguration(
            tenant_id=service_principal_config.get('tenant_id'),
            client_id=service_principal_config.get('client_id'),
            credential=service_principal_config.get('credential'),
            batch_account_resource_id=service_principal_config.get(
                'batch_account_resource_id'),
            storage_account_resource_id=service_principal_config.get(
                'storage_account_resource_id'),
        )

    shared_key_config = secrets_config.get('shared_key')
    batch = secrets_config.get('batch')
    storage = secrets_config.get('storage')

    if shared_key_config and (batch or storage):
        raise aztk.error.AztkError(
            "Shared keys must be configured either under 'sharedKey:' or under 'batch:' and 'storage:', not both."
        )

    if shared_key_config:
        secrets.shared_key = SharedKeyConfiguration(
            batch_account_name=shared_key_config.get('batch_account_name'),
            batch_account_key=shared_key_config.get('batch_account_key'),
            batch_service_url=shared_key_config.get('batch_service_url'),
            storage_account_name=shared_key_config.get('storage_account_name'),
            storage_account_key=shared_key_config.get('storage_account_key'),
            storage_account_suffix=shared_key_config.get(
                'storage_account_suffix'),
        )
    elif batch or storage:
        secrets.shared_key = SharedKeyConfiguration()
        if batch:
            log.warning(
                "Your secrets.yaml format is deprecated. To use shared key authentication use the shared_key key. See config/secrets.yaml.template"
            )
            secrets.shared_key.batch_account_name = batch.get(
                'batchaccountname')
            secrets.shared_key.batch_account_key = batch.get('batchaccountkey')
            secrets.shared_key.batch_service_url = batch.get('batchserviceurl')

        if storage:
            secrets.shared_key.storage_account_name = storage.get(
                'storageaccountname')
            secrets.shared_key.storage_account_key = storage.get(
                'storageaccountkey')
            secrets.shared_key.storage_account_suffix = storage.get(
                'storageaccountsuffix')

    docker_config = secrets_config.get('docker')
    if docker_config:
        secrets.docker = DockerConfiguration(
            endpoint=docker_config.get('endpoint'),
            username=docker_config.get('username'),
            password=docker_config.get('password'),
        )

    default_config = secrets_config.get('default')
    # Check for ssh keys if they are provided
    if default_config:
        secrets.ssh_priv_key = default_config.get('ssh_priv_key')
        secrets.ssh_pub_key = default_config.get('ssh_pub_key')


def read_cluster_config(
        path: str = aztk.utils.constants.DEFAULT_CLUSTER_CONFIG_PATH
) -> ClusterConfiguration:
    """
        Reads the config file in the .aztk/ directory (.aztk/cluster.yaml)
    """
    if not os.path.isfile(path):
        return

    with open(path, 'r') as stream:
        try:
            config_dict = yaml.load(stream)
        except yaml.YAMLError as err:
            raise aztk.error.AztkError(
                "Error in cluster.yaml: {0}".format(err))

        if config_dict is None:
            return

        return cluster_config_from_dict(config_dict)


def cluster_config_from_dict(config: dict):
    output = ClusterConfiguration()
    wait = False
    if config.get('id') is not None:
        output.cluster_id = config['id']

    if config.get('vm_size') is not None:
        output.vm_size = config['vm_size']

    if config.get('size'):
        output.vm_count = config['size']
        output.vm_low_pri_count = 0

    if config.get('size_low_pri'):
        output.vm_low_pri_count = config['size_low_pri']
        output.vm_count = 0

    if config.get('subnet_id') is not None:
        output.subnet_id = config['subnet_id']

    if config.get('username') is not None:
        output.user_configuration = UserConfiguration(
            username=config['username'])

        if config.get('password') is not None:
            output.user_configuration.password = config['password']

    if config.get('custom_scripts') not in [[None], None]:
        output.custom_scripts = []
        for custom_script in config['custom_scripts']:
            output.custom_scripts.append(
                aztk.spark.models.CustomScript(
                    script=custom_script['script'],
                    run_on=custom_script['runOn']))

    if config.get('azure_files') not in [[None], None]:
        output.file_shares = []
        for file_share in config['azure_files']:
            output.file_shares.append(
                aztk.spark.models.FileShare(
                    storage_account_name=file_share['storage_account_name'],
                    storage_account_key=file_share['storage_account_key'],
                    file_share_path=file_share['file_share_path'],
                    mount_path=file_share['mount_path'],
                ))

    if config.get('docker_repo') is not None:
        output.docker_repo = config['docker_repo']

    if config.get('plugins') not in [[None], None]:
        output.plugins = []
        for plugin in config['plugins']:
            ref = PluginReference.from_dict(plugin)
            output.plugins.append(ref.get_plugin())

    if config.get('worker_on_master') is not None:
        output.worker_on_master = config['worker_on_master']

    if config.get('wait') is not None:
        wait = config['wait']

    return output, wait


class SshConfig:
    def __init__(self):
        self.username = None
        self.cluster_id = None
        self.host = False
        self.connect = True

        # Set up ports with default values
        self.job_ui_port = '4040'
        self.job_history_ui_port = '18080'
        self.web_ui_port = '8080'

    def _read_config_file(
            self, path: str = aztk.utils.constants.DEFAULT_SSH_CONFIG_PATH):
        """
            Reads the config file in the .aztk/ directory (.aztk/ssh.yaml)
        """
        if not os.path.isfile(path):
            return

        with open(path, 'r') as stream:
            try:
                config = yaml.load(stream)
            except yaml.YAMLError as err:
                raise aztk.error.AztkError(
                    "Error in ssh.yaml: {0}".format(err))

            if config is None:
                return

            self._merge_dict(config)

    def _merge_dict(self, config):
        if config.get('username') is not None:
            self.username = config['username']

        if config.get('cluster_id') is not None:
            self.cluster_id = config['cluster_id']

        if config.get('job_ui_port') is not None:
            self.job_ui_port = config['job_ui_port']

        if config.get('job_history_ui_port') is not None:
            self.job_history_ui_port = config['job_history_ui_port']

        if config.get('web_ui_port') is not None:
            self.web_ui_port = config['web_ui_port']

        if config.get('jupyter_port') is not None:
            self.jupyter_port = config['jupyter_port']

        if config.get('name_node_ui_port') is not None:
            self.name_node_ui_port = config['name_node_ui_port']

        if config.get('rstudio_server_port') is not None:
            self.rstudio_server_port = config['rstudio_server_port']

        if config.get('host') is not None:
            self.host = config['host']

        if config.get('connect') is not None:
            self.connect = config['connect']

    def merge(self, cluster_id, username, job_ui_port, job_history_ui_port,
              web_ui_port, jupyter_port, name_node_ui_port,
              rstudio_server_port, host, connect):
        """
            Merges fields with args object
        """
        self._read_config_file(
            os.path.join(aztk.utils.constants.HOME_DIRECTORY_PATH, '.aztk',
                         'ssh.yaml'))
        self._read_config_file()
        self._merge_dict(
            dict(
                cluster_id=cluster_id,
                username=username,
                job_ui_port=job_ui_port,
                job_history_ui_port=job_history_ui_port,
                web_ui_port=web_ui_port,
                jupyter_port=jupyter_port,
                name_node_ui_port=name_node_ui_port,
                rstudio_server_port=rstudio_server_port,
                host=host,
                connect=connect))

        if self.cluster_id is None:
            raise aztk.error.AztkError(
                "Please supply an id for the cluster either in the ssh.yaml configuration file or with a parameter (--id)"
            )

        if self.username is None:
            raise aztk.error.AztkError(
                "Please supply a username either in the ssh.yaml configuration file or with a parameter (--username)"
            )


class JobConfig():
    def __init__(self):
        self.id = None
        self.applications = []
        self.custom_scripts = None
        self.spark_configuration = None
        self.vm_size = None
        self.docker_repo = None
        self.max_dedicated_nodes = None
        self.max_low_pri_nodes = None
        self.spark_defaults_conf = None
        self.spark_env_sh = None
        self.core_site_xml = None
        self.subnet_id = None
        self.worker_on_master = None

    def _merge_dict(self, config):
        config = config.get('job')

        if config.get('id') is not None:
            self.id = config['id']

        cluster_configuration = config.get('cluster_configuration')
        if cluster_configuration:
            self.vm_size = cluster_configuration.get('vm_size')
            self.docker_repo = cluster_configuration.get('docker_repo')
            self.max_dedicated_nodes = cluster_configuration.get('size')
            self.max_low_pri_nodes = cluster_configuration.get('size_low_pri')
            self.custom_scripts = cluster_configuration.get('custom_scripts')
            self.subnet_id = cluster_configuration.get('subnet_id')
            self.worker_on_master = cluster_configuration.get("worker_on_master")

        self.applications = config.get('applications')

        spark_configuration = config.get('spark_configuration')
        if spark_configuration:
            self.spark_defaults_conf = self.__convert_to_path(spark_configuration.get('spark_defaults_conf'))
            self.spark_env_sh = self.__convert_to_path(spark_configuration.get('spark_env_sh'))
            self.core_site_xml = self.__convert_to_path(spark_configuration.get('core_site_xml'))
            self.jars = [self.__convert_to_path(jar) for jar in spark_configuration.get('jars') or []]

    def __convert_to_path(self, str_path):
        if str_path:
            abs_path = os.path.abspath(os.path.expanduser(str_path))
            if not os.path.exists(abs_path):
                raise aztk.error.AztkError(
                    "Could not find file: {0}\nCheck your configuration file".
                    format(str_path))
            return abs_path

    def _read_config_file(
            self, path: str = aztk.utils.constants.DEFAULT_SPARK_JOB_CONFIG):
        """
            Reads the Job config file in the .aztk/ directory (.aztk/job.yaml)
        """
        if not path or not os.path.isfile(path):
            return

        with open(path, 'r') as stream:
            try:
                config = yaml.load(stream)
            except yaml.YAMLError as err:
                raise aztk.error.AztkError(
                    "Error in job.yaml: {0}".format(err))

            if config is None:
                return

            self._merge_dict(config)

    def merge(self, id, job_config_yaml=None):
        self._read_config_file(aztk.utils.constants.GLOBAL_SPARK_JOB_CONFIG)
        self._read_config_file(aztk.utils.constants.DEFAULT_SPARK_JOB_CONFIG)
        self._read_config_file(job_config_yaml)
        if id:
            self.id = id

        for entry in self.applications:
            if entry['name'] is None:
                raise aztk.error.AztkError(
                    "Application specified with no name. Please verify your configuration in job.yaml"
                )
            if entry['application'] is None:
                raise aztk.error.AztkError(
                    "No path to application specified for {} in job.yaml".
                    format(entry['name']))


def get_file_if_exists(file):
    local_conf_file = os.path.join(
        aztk.utils.constants.DEFAULT_SPARK_CONF_SOURCE, file)
    global_conf_file = os.path.join(aztk.utils.constants.GLOBAL_CONFIG_PATH,
                                    file)

    if os.path.exists(local_conf_file):
        return local_conf_file
    if os.path.exists(global_conf_file):
        return global_conf_file

    return None


def load_aztk_spark_config():
    return aztk.spark.models.SparkConfiguration(
        spark_defaults_conf=get_file_if_exists('spark-defaults.conf'),
        jars=load_jars(),
        spark_env_sh=get_file_if_exists('spark-env.sh'),
        core_site_xml=get_file_if_exists('core-site.xml'))


def load_jars():
    jars = None

    # try load global
    try:
        jars_src = os.path.join(aztk.utils.constants.GLOBAL_CONFIG_PATH,
                                'jars')
        jars = [os.path.join(jars_src, jar) for jar in os.listdir(jars_src)]
    except FileNotFoundError:
        pass

    # try load local, overwrite if found
    try:
        jars_src = os.path.join(aztk.utils.constants.DEFAULT_SPARK_CONF_SOURCE,
                                'jars')
        jars = [os.path.join(jars_src, jar) for jar in os.listdir(jars_src)]
    except FileNotFoundError:
        pass

    return jars
